### ISSUE TITLE
Grammer gives an error when the program crosses pages in the flash.

### ISSUE THREAD

**Issue by Zeda** [Zeda/Grammer2#15]
Hans Burch has reported this via email and they said that they have a modification that works. I'll be looking into this!

**Comment by NonstickAtom785** [comment]
Does Hans Burch know a lot about Grammer? Also wasn't this the error that happened with Archiving.

> Okay, it turned out that `bcall(_Arc_Unarc)` was overwriting code in RAM that handled page swapping (Grammer is a 2-page app, so some routines are on different pages. In my example, it never needed to go on the other page).

Wait just realized it was just overwriting code in RAM. Still is it similar?

**Comment by Zeda** [comment]
They are actually the reason that Grammer was revived! And they even worked on a modified version that fixes some bugs from the old version.

Basically, I lost the source code five or six years ago due to my computer failing, but they took a binary, disassembled it, then compared it to the most recent source available to math things up. Then they emailed me the source code and a year later I had it up on GitHub, with updates. They even documented some of the weird parts.

If I had to guess, I'd assume this bug is an edge case of the ReadArc routine I made. I tried being clever; maybe I was too clever :P

**Comment by Zeda** [comment]
This has been fixed! The issue was that I erroneously used `add hl,bc` when I needed to do `adc hl,bc`. Why? Because `add hl,bc` doesn't affect the sign flag, whereas `adc hl,bc` does, and I was using the sign flag to determine if the routine needed to cross a page boundary. Oof.

(For reference, these calcs use 16-bit addresses, so only 65536 bytes can be addressed. To get around the limitation, we have four 16384-byte "memory banks" and a port controls which 16384-byte flash or RAM page is mapped to a specific bank. Variables can cross pages, so when reading them into RAM, I need to put their starting page in the mem bank, read any data on that page, and if there is more to read on the next page, swap it out. Now the confounding factor is that Apps are stored in flash, so I need to temporarily swap the app itself out of the memory range! It gets complicated.)
